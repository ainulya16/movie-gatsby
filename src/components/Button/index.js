import styled from 'styled-components';
import { Button } from '@aia-digital/ui-library';

export default styled(Button)`
  width: ${({ theme }) => theme.color.w9};
  font: ${({ theme }) => theme.color.f9};
  color: #fff;
  border-radius: 5px;
  background: rgb(2, 0, 36);
  background: linear-gradient(
    180deg,
    rgba(2, 0, 36, 1) 0%,
    rgba(233, 24, 80, 1) 0%,
    rgba(200, 11, 62, 1) 100%
  );
  height: 50px !important;
  font-weight: bold;
  max-width: unset !important;
  display: inline-block;
  margin: auto;
`;
