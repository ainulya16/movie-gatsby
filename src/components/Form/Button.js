import React from 'react';
import { Button as ButtonDef, ButtonIcon } from '@aia-digital/ui-library';
import styled from 'styled-components';
import _ from 'lodash';
import { navigate } from '@reach/router';
import { Map, fromJS } from 'immutable';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { objectOf, any, string, instanceOf, func } from 'prop-types';
import API from '../../services/api';
import buildAPICALL from '../../helpers/buildCallAPI';
import { setOrderData } from '../../redux/Domain/action';
import { selectFormValues } from '../../redux/Form/selectors';
import selectDomain from '../../redux/Domain/selectors';

const CustomBodyList = {
  ORDERDATA: 'ORDERDATA',
};

const CustomCallbackList = {
  SETORDERDATA: 'SETORDERDATA',
};

const ButtonStyled = styled(ButtonDef)`
  width: ${({ theme }) => theme.color.w9};
  /* font: ${({ theme }) => theme.color.f9}; */
  font-size: 16px;
  border-radius: 4px;
  height: 50px;
  @media (max-width: 768px) {
    margin-bottom: 10px;
  }
`;
ButtonStyled.displayName = 'ButtonStyled';

const Button = ({
  label,
  customize,
  formdata,
  clicknext,
  clickprev,
  domainData,
  formReducerData,
  onSetOrderData,
}) => {
  const { action, componentProps, componentStyles } = customize;
  const domainValue = domainData.toJS();
  const formReducerDataValue = formReducerData.toJS();
  formdata = formdata ? formdata.toJS() : {};
  const source = {
    form: {
      ...formdata,
    },
  };

  const defaultActionFn = () => {};

  const oneMapCallBack = (response, target, callBack) => {
    const result = response.results[0] ? response.results[0] : {};
    if (result && !_.isEmpty(result)) {
      for (let index = 0; index < target.length; index++) {
        const element = target[index];
        callBack('application', element.field, result[element.source]);
      }
    }
  };

  const actionApiCall = () => {
    const callBackSource = [
      {
        service: 'https://developers.onemap.sg',
        fn: oneMapCallBack,
      },
    ];
    const callBack = action.fn;
    buildAPICALL(
      action.service,
      action.api,
      action.method || 'get',
      action.body || {},
      source || {},
    ).then(res => {
      const service = _.find(callBackSource, x => x.service === action.service);
      service.fn(res, action.customize.target, callBack);
    });
  };

  const bodyResolver = body => {
    const { ORDERDATA } = CustomBodyList;
    const orderData = {
      ...domainValue,
      application: _.get(formReducerDataValue, 'values', {}),
    };

    if (body.type === 'custom') {
      // for custom body
      switch (body.value) {
        case ORDERDATA:
          return orderData;

        default:
          return body.value;
      }
    }
    return body.value ? body.value : body;
  };

  const apiCall = async () => {
    const { path, url, method, body, endpoint } = action;
    const { SETORDERDATA } = CustomCallbackList;
    if (!path || !url || !method || !body || !endpoint) {
      console.log('Request is not valid');
    }
    const api = new API(endpoint);
    clicknext()
      .then(async () => {
        if (method === 'post') {
          const data = { path, method, url, body: bodyResolver(body) };
          try {
            const res = await api.post(data);
            const onSetOrderDataCB = _.find(
              action.customize.callback,
              cb => cb === SETORDERDATA,
            );
            // const onSetDomainDataCB = _.find(action.customize.callback, cb => cb === "onSetDomainData");
            if (!_.isEmpty(onSetOrderDataCB)) {
              onSetOrderData(res.data.orderReference);
            }
            return res;
          } catch (error) {
            console.log('Error', error);
            return error;
          }
        }
      })
      .catch(err => {
        console.log('Error', err);
        return err;
      });
  };

  const backFn = () => {
    const { url } = action;
    if (url) navigate(url);
    clickprev();
  };

  const sourceFn = [
    {
      name: 'populate',
      fn: actionApiCall,
    },
    {
      name: 'clicknext',
      fn: clicknext,
    },
    {
      name: 'clickprevious',
      fn: backFn,
    },
    {
      name: 'apiCall',
      fn: apiCall,
    },
  ];

  const actionFn = () => {
    const btnAction = !_.isEmpty(action)
      ? _.find(sourceFn, x => x.name === action.name)
      : { fn: defaultActionFn };
    btnAction.fn();
  };

  // if (customize.action && customize.action.type === 'api') {
  //   actionFn = actionApiCall;
  // } else {
  //   actionFn = defaultActionFn;
  // }

  // if (fieldValue.type && fieldValue.type === 'static') {
  //   radioValue = fieldValue.values;
  // }
  return customize.type && customize.type === 'icon' ? (
    <ButtonIcon onClick={actionFn} iconName={label} circle size="m" shadow />
  ) : (
    <ButtonStyled
      style={componentStyles}
      onClick={actionFn}
      {...componentProps}
    >
      {label}
    </ButtonStyled>
  );
};

Button.propTypes = {
  label: string,
  customize: objectOf(any).isRequired,
  domainData: instanceOf(Map),
  formReducerData: instanceOf(Map),
  formdata: instanceOf(Map),
  onSetOrderData: func,
  clicknext: func,
  clickprev: func,
};

Button.defaultProps = {
  label: 'Ok',
  domainData: fromJS({}),
  formReducerData: fromJS({}),
  formdata: fromJS({}),
  onSetOrderData: () => {},
  clicknext: () => {},
  clickprev: () => {},
};

export const mapStateToProps = createStructuredSelector({
  formReducerData: selectFormValues('application'),
  domainData: selectDomain(),
});

const mapDispatchToProps = {
  onSetOrderData: setOrderData,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Button);
