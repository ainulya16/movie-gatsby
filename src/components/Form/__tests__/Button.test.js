import React from 'react';
import { shallow } from 'enzyme';
// import { Field } from 'redux-form/immutable';
import renderer from 'react-test-renderer';
import Button from '../Button';

describe('Button Component', () => {
  //   it('should have redux-form Field', () => {
  //     const component = shallow(<Button name="test" />);
  //     expect(component.find(Field).exists()).toBe(true);
  //   });
  it('should be exists', () => {
    const mockLabel = 'MOCK LABEL';
    const mockCustomize = {
      action: {
        name: 'populate',
        customize: {
          target: [
            {
              field: 'postalCodeMailing',
              source: 'POSTAL',
            },
            {
              field: 'streetMailing',
              source: 'ADDRESS',
            },
            {
              field: 'blockMailing',
              source: 'BLK_NO',
            },
            {
              field: 'buildingMailing',
              source: 'BUILDING',
            },
          ],
        },
        type: 'api',
        service: 'https://developers.onemap.sg',
        method: 'get',
        api: 'commonapi/search',
        body: {
          searchVal: '{{form.application.values.postalCodeMailing}}',
          returnGeom: 'Y',
          getAddrDetails: 'Y',
        },
      },
    };
    const wrapper = shallow(
      <Button label={mockLabel} customize={mockCustomize} />,
    );
    expect(wrapper.exists()).toBe(true);
  });
  it('should be exists when action type is not api', () => {
    const mockLabel = 'MOCK LABEL';
    const mockCustomize = {
      action: {
        type: 'other',
      },
    };
    const wrapper = shallow(
      <Button label={mockLabel} customize={mockCustomize} />,
    );
    expect(wrapper.exists()).toBe(true);
  });
  it('should match snapshot', () => {
    const mockLabel = 'MOCK LABEL';
    const mockCustomize = {
      action: {
        type: 'api',
      },
    };
    const tree = renderer.create(
      <Button label={mockLabel} customize={mockCustomize} />,
    );
    expect(tree).toMatchSnapshot();
  });
  it('should call defaultActionFn when button clicked', () => {
    const mockLabel = 'MOCK LABEL';
    const mockCustomize = {};
    const wrapper = shallow(
      <Button label={mockLabel} customize={mockCustomize} />,
    );
    wrapper.find('ButtonStyled').simulate('click');
  });
  it('should call defaultActionFn when button clicked', () => {
    const mockLabel = 'MOCK LABEL';
    const mockCustomize = {
      action: {
        type: 'api',
      },
    };
    const wrapper = shallow(
      <Button label={mockLabel} customize={mockCustomize} />,
    );
    wrapper.find('ButtonStyled').simulate('click');
  });
});
