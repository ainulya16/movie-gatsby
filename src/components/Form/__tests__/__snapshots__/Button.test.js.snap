// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should match snapshot 1`] = `
.c0 {
  min-width: 250px;
  width: 250px;
  text-align: center;
}

.c2 {
  height: 45px;
  font-family: 'AIAMedium';
  text-transform: none;
  font-size: 1.1em;
  border: 0px;
  display: inline-block;
  background: #D31145;
  -webkit-transition: width 0.3s ease,min-width 0.3s ease,background-color 0.3s ease,border-radius 0.3s ease,border 0.3s ease,display 0.3s ease;
  transition: width 0.3s ease,min-width 0.3s ease,background-color 0.3s ease,border-radius 0.3s ease,border 0.3s ease,display 0.3s ease;
  color: #FFFFFF;
  background-color: #FFF;
  background-image: linear-gradient(to bottom,#d31145,#D31145,#d31145);
  border: none;
}

.c2:active {
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
  -webkit-transition: scale 0.1s ease-in-out;
  transition: scale 0.1s ease-in-out;
}

.c2:focus {
  outline: none;
}

.c2:disabled {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 0.4;
  -webkit-filter: alpha(opacity=40);
  filter: alpha(opacity=40);
}

.c1 {
  width: 90%;
  min-width: unset;
  max-width: 250px;
}

@media (max-width:768px) {
  .c1 {
    margin-bottom: 10px;
  }
}

@media (min-width:768px) {
  .c1 {
    margin-top: 40px;
  }
}

<div
  className="c0"
  style={Object {}}
>
  <button
    className="btn c1 c2"
    color="#FFFFFF"
    disabled={false}
    onClick={[Function]}
    radius={null}
  >
    <div
      className="child"
    >
      MOCK LABEL
    </div>
  </button>
</div>
`;
